gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java

```diff
@@ -1095,6 +1117,26 @@ boolean notRequiredByAnyTx( LocalRegion region, RegionEntry re,
 
   }
 
+  static class QueueRemover extends
+      MapCallbackAdapter<Object, Object, Object, Object> {
+
+    @Override
+    public Object removeValue(Object key, Object value, Object existingValue,
+        Object context, Object removeParams) {
+      if (value != null
+          && (value == NO_OBJECT_TOKEN || ((existingValue == value)
+          && ((BlockingQueue)existingValue).size() == 0))) {
```

Comment:
linked queue size() implementation might not be efficient so always better to use isEmpty()